embrace-gradle-plugin/src/main/java/io/embrace/android/gradle/plugin/tasks/ndk/UploadSharedObjectFilesTask.kt

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * This task handles the upload of shared object (.so) files to the Embrace backend.
+ *
+ * Its most important inputs are:
+ * - compressedSharedObjectFilesDirectory: Directory containing the compressed .so files.
+ * - architecturesToHashedSharedObjectFilesMapJson: JSON mapping architectures to file hashes.
+ *
+ * Given those inputs, the task will:
+ * - Report the available shared object files to the backend.
+ * - Find the shared object files requested by the backend (if any) in the compressed files directory.
+ * - Upload the requested files individually to the backend.
+ */
+abstract class UploadSharedObjectFilesTask @Inject constructor(
+    objectFactory: ObjectFactory,
+) : EmbraceUploadTask, EmbraceUploadTaskImpl(objectFactory) {
+
+    private val serializer = MoshiSerializer()
+
+    @get:Input
+    val failBuildOnUploadErrors: Property<Boolean> = objectFactory.property(Boolean::class.java)
+
+    @SkipWhenEmpty
+    @get:InputDirectory
+    val compressedSharedObjectFilesDirectory: DirectoryProperty = objectFactory.directoryProperty()
+
+    @SkipWhenEmpty
+    @get:InputFile
+    val architecturesToHashedSharedObjectFilesMapJson: RegularFileProperty = objectFactory.fileProperty()
+
+    private lateinit var architecturesToHashedSharedObjectFilesMap: Map<String, Map<String, String>>
+
+    private val okHttpNetworkService by lazy {
+        OkHttpNetworkService(requestParams.get().baseUrl)
+    }
+
+    @TaskAction
+    fun onRun() {
+        architecturesToHashedSharedObjectFilesMap = getArchToFilenameToHashMap()
+        val requestedSharedObjectFiles = getRequestedSharedObjectFiles() ?: return
+        val foundRequestedSharedObjectFiles = findRequestedSharedObjectFiles(requestedSharedObjectFiles) ?: return
+        uploadSharedObjectFiles(foundRequestedSharedObjectFiles)
+    }
+
+    private fun getArchToFilenameToHashMap() = try {
+        serializer.fromJson(
+            architecturesToHashedSharedObjectFilesMapJson.get().asFile.bufferedReader().use { it.readText() },
+            ArchitecturesToHashedSharedObjectFilesMap::class.java
+        ).architecturesToHashedSharedObjectFiles
+    } catch (exception: Exception) {
+        error("Failed to read the architectures to hashed shared object files map: ${exception.message}")
+    }
+
+    /**
+     * Report the files we have available, and let the backend tell us which ones it needs
+     *
+     * @return a map of architectures to requested symbols
+     */
+    private fun getRequestedSharedObjectFiles(): Map<String, List<String>>? {
+        val params = NdkUploadHandshakeRequest(
+            requestParams.get().appId,
+            requestParams.get().apiToken,
+            variantData.get().name,
+            architecturesToHashedSharedObjectFilesMap
+        )
+
+        // TODO: we are swallowing backend errors, we should fix this.
+        return NdkUploadHandshake(okHttpNetworkService).getRequestedSymbols(params, failBuildOnUploadErrors.get())
```

Comment:
Is this comment still valid? Can we just propagate whatever error/exception is getting thrown and fail if that flag is enabled?

Reply:
It is. If the backend returns an error, getRequestedSymbols returns null. We don't do anything in particular for that case.

Reply:
I plan to review all of the TODOs when I start writing tests and reviewing edge cases tomorrow